design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+
+This is not a final proposal, but rather a lens to explore both unixfsv2
+concepts and schemas themselves at the same time, and how we expect them
+to contribute to each other.  Several concepts in schemas, advanced layouts,
+their interactions, and their syntaxes will be trialled here; they are not final.
+
+Without further ado...
+
+Schema
+------
+
+```
+type AnyFile union {
+	| File "f"
+	| Dir "d"
+	| Symlink "s"
+} representation keyed
+
+type Filename string
+
+type Dir struct {
+	members {Filename:DirEnt}<HAMT>
+}
+
+type DirEnt struct {
+	content &AnyFile
+	attribs Attribs
+}
+
+type File struct {
+	content Bytes<Rabin>
+}
+
+type Symlink struct {
+	target String
+}
+
+advanced Rabin {
+	kind bytes
+	implementation "IPLD/experimental/rabin/v1"
+}
+
+advanced HAMT {
+	kind map
+	implementation "IPLD/experimental/HAMT/v1"
+}
+
+type Attribs struct {
+	mtime Int # Danger danger will robinson; int sizes will rear their ugly head.
+	posix Int # Can we just... do the standard 0777 mask packing here?  I don't think anything else is *less* work.
+	sticky Bool (implicit: false)
+	setuid Bool (implicit: false)
+	setgid Bool (implicit: false)
+	uid Int # make another schema which deliberately excludes this.  try to match this one (the one with more info) first.
+	gid Int # make another schema which deliberately excludes this.  try to match this one (the one with more info) first.
+}
+
+# You should be able to imagine sending a fancy graphsync request with some structure like:
+#
+#  {"schema": {{schemaCIDofUnixfs}}, "selector": {{selectorSpec}},
+#   "schemaPatch": {"./dir/transparently/through/to/file": {{schemaCIDofRabinLayout}},
+#   "startAt": {{rootBlockCID+path}} }
+#
+# This kind of fancy graphsync request can be used to make a query which does "the right thing" with
+# all advanced layouts transparently; then when reaching the 'patch' area, perhaps the selector spec
+# wants to use the seeded-psuedorandomness feature, so it can issue queries like this across many peers.
+#
+# Perhaps we won't actually need the 'patch' feature, because in practice if doing queries with lots
+# of fancy sharded requests, prooobably you'll do the path *up to* that first, in a separate query,
+# But either way, you can see how using schemas for signaling is useful here.
+```
+
+Interesting Highlights
+----------------------
+
+### Using various schemas as a feature!
+
+Observe the comments on the `Attribs.uid` and `Attribs.gid` fields.
+
+By using several different (but significantly overlapping) schemas to describe variations
+in which attributes we're storing, we can make unixfsv2 into an extensible yet still
+very well defined system.
+
+Concretely: If the `uid` and `gid` elements are found present in the first `Attribs` map
+processed in a tree of data, then we match the first schema (the one with more info),
+and we can now expect *all future* `Attribs` maps to consistently include the same data.
```

Comment:
I’m not sure I understand what “all future” means in this context. Do you mean all future sub-files and directories in the tree?